src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java

```diff
@@ -219,23 +220,50 @@ public boolean fullyGeometricallyEncloses(final Location location)
      *            A {@link PolyLine} item
      * @return True if the {@link MultiPolygon} contains the provided {@link PolyLine}.
      */
+    @Override
     public boolean fullyGeometricallyEncloses(final PolyLine polyLine)
     {
-        for (final Polygon inner : inners())
+        for (final Polygon outer : outers())
         {
-            if (inner.overlaps(polyLine))
+            if (outer.fullyGeometricallyEncloses(polyLine))
             {
-                return false;
+                return this.getOuterToInners().get(outer).stream()
+                        .noneMatch(inner -> inner.overlaps(polyLine));
             }
         }
-        for (final Polygon outer : outers())
+        return false;
+    }
+
+    /**
+     * Tests to see if entire surface of the provided {@link MultiPolygon} lies within this
+     * {@link MultiPolygon}
+     * 
+     * @param that
+     *            the provided {@link MultiPolygon} to test
+     * @return true if the conditions are met, false otherwise
+     */
+    @Override
+    public boolean fullyGeometricallyEncloses(final MultiPolygon that)
+    {
+        final QuadTree<Polygon> thisOuters = QuadTree.forLocated(this.outers());
```

Comment:
`QuadTree` is usually a good choice for Point objects. For anything with a shape, I would recommend using an `RTree` instead.

Reply:
Did you test the performance on this one out of curiosity?

Reply:
I didn't take any time statistics. Are there two different options you would like me to compare the performance of?